macros/generate_surrogate_key.sql

```diff
@@ -0,0 +1,31 @@
+
+{# The original implementation casted field_list elements to LONG VARCHAR. This led to #}
+{# 'A column or character expression is larger than the max size.' error. This #}
+{# implementation removes the cast. It is the responsibility of the caller to make sure #}
+{# that provided field names are strings #}
+
+{%- macro teradata__generate_surrogate_key(field_list) -%}
+
+{%- if var('surrogate_key_treat_nulls_as_empty_strings', False) -%}
+    {%- set default_null_value = "" -%}
+{%- else -%}
+    {%- set default_null_value = '_dbt_utils_surrogate_key_null_' -%}
+{%- endif -%}
+
+{%- set fields = [] -%}
+
+{%- for field in field_list -%}
+
+    {%- do fields.append(
+        "coalesce(" ~ field ~ ", '" ~ default_null_value  ~"')"
```

Comment:
Are we forcing the customers to mention the field names as strings?

Reply:
I'm sorry I do not understand your question. generate_surrogate_key requires a list values to be used for the surrogate key calculation. Our implementation required this list to be contain only strings, because of the mentioned problem in the header of the file. Every null value from the list is converted to default_null_value string

> {# The original implementation casted field_list elements to LONG VARCHAR. This led to #}
> {# 'A column or character expression is larger than the max size.' error. This #}